### PR TITLE
`azurerm_kubernetes_cluster``azurerm_kubernetes_cluster_node_pool` - support for the `node_public_ip_tags` property

### LINKED ISSUE #19128: Support for internet routing ip tag on node level public ip in AKS
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

There is no option to specify the ip tag on the node level public ip's in AKS nodepool. This is needed to override the routing preference on the node level public ip to internet. 

### New or Affected Resource(s)/Data Source(s)

azurerm_3.29.1 and below

### Potential Terraform Configuration

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by ms-henglu**
Closes #19128

### CODE REVIEW

File: internal/services/containers/kubernetes_nodepool.go [hunk 1]
```diff
@@ -1655,3 +1698,56 @@ func expandClusterNodePoolUpgradeSettings(input []interface{}) *managedclusters.
...
+}
+
+func flattenClusterPoolNetworkProfile(input *managedclusters.AgentPoolNetworkProfile) []interface{} {
+	if input == nil {
+		return []interface{}{}
+	}
+
+	return []interface{}{
+		map[string]interface{}{
+			"node_public_ip_tags": flattenClusterPoolNetworkProfileNodePublicIPTags(input.NodePublicIPTags),
```
Comment (stephybun) [inline internal/services/containers/kubernetes_nodepool.go:1735]: This should only be returned if there are actually tags set/returned by the API

File: internal/services/containers/kubernetes_cluster_node_pool_resource_test.go [hunk 2]
```diff
@@ -2442,3 +2457,43 @@ resource "azurerm_kubernetes_cluster_node_pool" "test" {
...
+}
+
+resource "azurerm_kubernetes_cluster_node_pool" "test" {
+  name                  = "internal"
+  kubernetes_cluster_id = azurerm_kubernetes_cluster.test.id
+  vm_size               = "Standard_D2s_v3"
+  enable_node_public_ip = true
+  network_profile {
+    node_public_ip_tags = {
+      RoutingPreference = "Internet"
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_node_pool_resource_test.go:2494]: Can we set multiple tags here

File: internal/services/containers/kubernetes_cluster_network_resource_test.go [hunk 3]
```diff
@@ -3214,3 +3229,37 @@ resource "azurerm_kubernetes_cluster" "test" {
...
+  resource_group_name = azurerm_resource_group.test.name
+  dns_prefix          = "acctestaks%[2]d"
+  default_node_pool {
+    name                  = "default"
+    node_count            = 1
+    vm_size               = "Standard_DS2_v2"
+    enable_node_public_ip = true
+    network_profile {
+      node_public_ip_tags = {
+        RoutingPreference = "Internet"
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_network_resource_test.go:3256]: Same here

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks for this PR @ms-henglu. I left a few minor comments in-line, but once those are fixed up this should be good to go!

File: internal/services/containers/kubernetes_cluster_node_pool_resource_test.go (hunk 2 above)
Comment (ms-henglu) [reply internal/services/containers/kubernetes_cluster_node_pool_resource_test.go:2494]: Hi @stephybun , I've confirmed with the feature owner, that now it only supports one ip tag, but will support multiple in the future.

**Review by stephybun — CHANGES_REQUESTED** [review]

**Review by stephybun — CHANGES_REQUESTED** [review]
There are test failures:
```
------- Stdout: -------
=== RUN   TestAccKubernetesClusterNodePool_nodeIPTags
=== PAUSE TestAccKubernetesClusterNodePool_nodeIPTags
=== CONT  TestAccKubernetesClusterNodePool_nodeIPTags
testcase.go:110: Step 1/2 error: After applying this test step, the plan was not empty.
stdout:
Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement
Terraform will perform the following actions:
# azurerm_kubernetes_cluster_node_pool.test must be replaced
-/+ resource "azurerm_kubernetes_cluster_node_pool" "test" {
- custom_ca_trust_enabled = false -> null
- enable_auto_scaling     = false -> null
- enable_host_encryption  = false -> null
- fips_enabled            = false -> null
~ id                      = "/subscriptions/*******/resourceGroups/acctestRG-aks-230119131734642943/providers/Microsoft.ContainerService/managedClusters/acctestaks230119131734642943/agentPools/internal" -> (known after apply)
~ kubelet_disk_type       = "OS" -> (known after apply)
- max_count               = 0 -> null
~ max_pods                = 110 -> (known after apply)
- min_count               = 0 -> null
name                    = "internal"
~ node_count              = 0 -> (known after apply)
~ node_labels             = {} -> (known after apply)
~ orchestrator_version    = "1.24.6" -> (known after apply)
~ os_disk_size_gb         = 128 -> (known after apply)
~ os_sku                  = "Ubuntu" -> (known after apply)
# (10 unchanged attributes hidden)
+ node_network_profile {
+ node_public_ip_tags = {
+ "RoutingPreference" = "Internet"
} # forces replacement
}
}
Plan: 1 to add, 0 to change, 1 to destroy.
--- FAIL: TestAccKubernetesClusterNodePool_nodeIPTags (872.74s)
FAIL
```
```
------- Stdout: -------
=== RUN   TestAccKubernetesCluster_clusterPoolNodePublicIPTags
=== PAUSE TestAccKubernetesCluster_clusterPoolNodePublicIPTags
=== CONT  TestAccKubernetesCluster_clusterPoolNodePublicIPTags
testcase.go:110: Step 1/2 error: After applying this test step, the plan was not empty.
stdout:
Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement
Terraform will perform the following actions:
# azurerm_kubernetes_cluster.test must be replaced
-/+ resource "azurerm_kubernetes_cluster" "test" {
~ api_server_authorized_ip_ranges     = [] -> (known after apply)
- enable_pod_security_policy          = false -> null
~ fqdn                                = "acctestaks230119130043638271-64f6eacc.hcp.eastus.azmk8s.io" -> (known after apply)
+ http_application_routing_zone_name  = (known after apply)
~ id                                  = "/subscriptions/*******/resourceGroups/acctestRG-aks-230119130043638271/providers/Microsoft.ContainerService/managedClusters/acctestaks230119130043638271" -> (known after apply)
~ kube_admin_config                   = (sensitive value)
+ kube_admin_config_raw               = (sensitive value)
~ kube_config                         = (sensitive value)
~ kube_config_raw                     = (sensitive value)
~ kubernetes_version                  = "1.24.6" -> (known after apply)
- local_account_disabled              = false -> null
name                                = "acctestaks230119130043638271"
~ node_resource_group                 = "MC_acctestRG-aks-230119130043638271_acctestaks230119130043638271_eastus" -> (known after apply)
- oidc_issuer_enabled                 = false -> null
+ oidc_issuer_url                     = (known after apply)
~ portal_fqdn                         = "acctestaks230119130043638271-64f6eacc.portal.hcp.eastus.azmk8s.io" -> (known after apply)
+ private_dns_zone_id                 = (known after apply)
+ private_fqdn                        = (known after apply)
# (12 unchanged attributes hidden)
+ api_server_access_profile {
+ authorized_ip_ranges     = (known after apply)
+ subnet_id                = (known after apply)
+ vnet_integration_enabled = (known after apply)
}
+ auto_scaler_profile {
+ balance_similar_node_groups      = (known after apply)
+ empty_bulk_delete_max            = (known after apply)
+ expander                         = (known after apply)
+ max_graceful_termination_sec     = (known after apply)
+ max_node_provisioning_time       = (known after apply)
+ max_unready_nodes                = (known after apply)
+ max_unready_percentage           = (known after apply)
+ new_pod_scale_up_delay           = (known after apply)
+ scale_down_delay_after_add       = (known after apply)
+ scale_down_delay_after_delete    = (known after apply)
+ scale_down_delay_after_failure   = (known after apply)
+ scale_down_unneeded              = (known after apply)
+ scale_down_unready               = (known after apply)
+ scale_down_utilization_threshold = (known after apply)
+ scan_interval                    = (known after apply)
+ skip_nodes_with_local_storage    = (known after apply)
+ skip_nodes_with_system_pods      = (known after apply)
}
~ default_node_pool {
- custom_ca_trust_enabled      = false -> null
- enable_auto_scaling          = false -> null
- enable_host_encryption       = false -> null
- fips_enabled                 = false -> null
~ kubelet_disk_type            = "OS" -> (known after apply)
- max_count                    = 0 -> null
~ max_pods                     = 110 -> (known after apply)
- min_count                    = 0 -> null
name                         = "default"
~ node_labels                  = {} -> (known after apply)
- only_critical_addons_enabled = false -> null
~ orchestrator_version         = "1.24.6" -> (known after apply)
~ os_disk_size_gb              = 128 -> (known after apply)
~ os_sku                       = "Ubuntu" -> (known after apply)
+ workload_runtime             = (known after apply)
# (7 unchanged attributes hidden)
+ node_network_profile {
+ node_public_ip_tags = {
+ "RoutingPreference" = "Internet"
} # forces replacement
}
}
~ identity {
~ principal_id = "302382d3-eb99-4e11-b83e-0e2d5ca4a104" -> (known after apply)
~ tenant_id    = "*******" -> (known after apply)
# (1 unchanged attribute hidden)
}
~ kubelet_identity {
~ client_id                 = "d63e049c-7ea2-4716-afd1-adf4269719f0" -> (known after apply)
~ object_id                 = "4bccf6e5-0892-4457-b60b-cec34641e670" -> (known after apply)
~ user_assigned_identity_id = "/subscriptions/*******/resourceGroups/MC_acctestRG-aks-230119130043638271_acctestaks230119130043638271_eastus/providers/Microsoft.ManagedIdentity/userAssignedIdentities/acctestaks230119130043638271-agentpool" -> (known after apply)
}
~ network_profile {
~ dns_service_ip      = "10.0.0.10" -> (known after apply)
~ docker_bridge_cidr  = "172.17.0.1/16" -> (known after apply)
+ ebpf_data_plane     = (known after apply)
~ ip_versions         = [
- "IPv4",
] -> (known after apply)
~ load_balancer_sku   = "standard" -> (known after apply)
+ network_mode        = (known after apply)
~ network_plugin      = "kubenet" -> (known after apply)
+ network_plugin_mode = (known after apply)
+ network_policy      = (known after apply)
~ outbound_type       = "loadBalancer" -> (known after apply)
~ pod_cidr            = "10.244.0.0/16" -> (known after apply)
~ pod_cidrs           = [
- "10.244.0.0/16",
] -> (known after apply)
~ service_cidr        = "10.0.0.0/16" -> (known after apply)
~ service_cidrs       = [
- "10.0.0.0/16",
] -> (known after apply)
~ load_balancer_profile {
~ effective_outbound_ips      = [
- "/subscriptions/*******/resourceGroups/MC_acctestRG-aks-230119130043638271_acctestaks230119130043638271_eastus/providers/Microsoft.Network/publicIPAddresses/28086d92-d5f8-4de2-842f-36a21059e706",
] -> (known after apply)
~ idle_timeout_in_minutes     = 0 -> (known after apply)
~ managed_outbound_ip_count   = 1 -> (known after apply)
~ managed_outbound_ipv6_count = 0 -> (known after apply)
~ outbound_ip_address_ids     = [] -> (known after apply)
~ outbound_ip_prefix_ids      = [] -> (known after apply)
~ outbound_ports_allocated    = 0 -> (known after apply)
}
+ nat_gateway_profile {
+ effective_outbound_ips    = (known after apply)
+ idle_timeout_in_minutes   = (known after apply)
+ managed_outbound_ip_count = (known after apply)
}
}
+ windows_profile {
+ admin_password = (sensitive value)
+ admin_username = (known after apply)
+ license        = (known after apply)
+ gmsa {
+ dns_server  = (known after apply)
+ root_domain = (known after apply)
}
}
}
Plan: 1 to add, 0 to change, 1 to destroy.
--- FAIL: TestAccKubernetesCluster_clusterPoolNodePublicIPTags (815.74s)
FAIL
```

**Comment by ms-henglu** [comment]
Hi @stephybun ,

Would you please take another look? The tests are passed.

**Review by stephybun — APPROVED** [review]
Thanks @ms-henglu. I reverted the last change which added a check on the length of the tags, since that would become problematic if more properties are added to the network profile block for nodes. For now this is fine.

LGTM 👍

**Pull request merged**